### ISSUE TITLE
React-Table's current state and future roadmap

### ISSUE THREAD

**Issue by snezik** [tannerlinsley/react-table#1143]
No updates lately.
There are a lot of pull requests, please merge some fixes and features.

Don't let the project die

**Comment by tannerlinsley** [comment]
Clearly this issue was meant as a request for an update on the state of React-Table, so let's do just that:

**TL;DR: React-Table is still being actively maintained, dog-fooded, and developed by it's original maintainers and avid users.** (Primarily me, @tannerlinsley, with a bit of help from @gary-menzel). As time permits, it is undergoing a remarkable rewrite as we speak to become a more sustainable collection of headless components, UI implementations and composable HOC's/plugins.

The unsustainable API of the current version has been wildly successful in helping people become immediately productive with their use-cases, but its limited flexibility and underlying architecture is to blame for the massive amount of pull requests and issues that come in every day. 

@gary-menzel, other maintainers and myself have all previously tried to keep up with this influx of feature requests and bugs, and all of us have experienced some level of burnout while we endlessly add more props and API surface area to the library. We have learned the hard way that this is no way to build or maintain a library with the scope as complex as React-Table's.

So, instead of putting out ever-growing fires and throwing more props at the problem, we have chosen to move forward by building a less "flammable" library 😄 using headless components as an escape hatch. If you are not familiar with headless components, you will benefit from reading this article: https://medium.com/merrickchristensen/headless-user-interface-components-565b0c0f2e18.

Headless components will unlock the ultimate level of flexibility by inverting most control back to the user. This will allow for customization that wasn't possible before with React-Table and also allow the library to become leaner, smarter, and more maintainable.

The first fruits of this effort will likely be publicly visible (a new version branch) in the next few months.

Tentative Roadmap:
- 3-5 Months: New version branch for headless components (partial-parity)
- 1-2 Months: UI implementations using these components (feature-parity)
- 1 Month: Migration guides, new features, deprecations, etc.
- 1 Month: Release

*RE: Original issue: "Wake Up" probably isn't the best issue title. Let's try to be more respectful in the future. No need to let React-Table slip into a demeaning place as a community.*

**Comment by tannerlinsley** [comment]
Under best case scenarios yes.

**Comment by anthonychung14** [comment]
+1 to @tannerlinsley I discovered this library 2-3 months ago and have been blown away by how thorough the documentation and examples were. Our team has gone all-in on React-Table thanks to you!

**Comment by kylehalleman** [comment]
@tannerlinsley Will a new release be cut and published to npm with the recent PRs that were merged into master?

**Comment by hansena** [comment]
@tannerlinsley are there any pieces of this effort that the community can contribute to?

**Comment by tannerlinsley** [comment]
The pieces are close to being published in a new branch. Stay tuned for a big todo list where you can help :)

**Comment by kurtwilbies** [comment]
@tannerlinsley

I suggest first publish a 6.9 version and then open the new one.

**Comment by rosibaj** [comment]
When will this release be completed? We are trying to decide how much accessibility effort we need to put where!  If its coming relatively quickly that would help :)

**Comment by zhDmitry** [comment]
@tannerlinsley any updates on this roadmap ?

**Comment by kurtwilbies** [comment]
@tannerlinsley Any updates?

**Comment by tannerlinsley** [comment]
V7 is still likely not to be out for a while. Keep using v6, fork if necessary, and maintain your list of features/upgrades until v7-alpha materializes.

**Comment by zhDmitry** [comment]
@tannerlinsley btw, can u provide general idea of v7 and how api could change ?

**Comment by tannerlinsley** [comment]
# Updates

- The initial alpha source code for v7 is now available on the master branch.
- Likewise, v6 has been moved to its own branch now. 
- I will not be actively working on any more v6 issues submitted through Github.
- I will, however, accept any non-breaking PR's to fix critical v6 bugs
- I am estimating another 2 months before v7 is ready for a stable release.

**Comment by Grsmto** [comment]
Any reason v7 is not released under pre-release tag?
As I understand there will be a huge change in the API so we can't start using React-Table right now until the stable release.
Would be great to have it on npm somewhere to use it!

**Comment by tannerlinsley** [comment]
It's available under react-table@next!

- The api is still changing a lot
- Things will break
- Use for experimenting/ideas/contributing only ;)

**Comment by mmazanec22** [comment]
@tannerlinsley congrats on the new release!  I saw your comments on several issues and PRs that accessibility compliance was part of the plan for v7.  Any word on that?

**Comment by tannerlinsley** [comment]
@mmazanec22 Accessibility will be the responsibility of the user, since v7 is a **headless** React hook. Thus, you can approach accessibility any way you'd like when you build your table's markup/renderer.

**Comment by labtocat** [comment]
@tannerlinsley How much of an effort will be there for a developer to migrate from v6 to v7? Will the code breaks?

**Comment by tannerlinsley** [comment]
I will be doing my best to write a migration guide, but there will be work involved. Close to all of that work will be building your table's markup, since v7 is headless. I'll be providing basic markup examples, along with some styled-components examples to replicate the current table very closely.

As far as table configuration, most things will translate over directly (props will become options in the `useReactTable` hook) and HOC's will either become components you use in your table markup, or just plugin-hooks.

Regardless, expect great documentation and at the very least a basic migration guide.

**Comment by Grsmto** [comment]
For anyone looking for this, I'm currently working on a React Table v7 TypeScript typings definitions: https://gist.github.com/Grsmto/1ac3a7ddb8ad8288660784a37bff1798
Plan is to create a PR to `definitely-typed` repository once the API of v7 is a bit more stable.
Ping me if anyone wants to help with this.

**Comment by gowthamkumarguruju** [comment]
anyone working on this issues https://github.com/tannerlinsley/react-table/issues/1079 ???

**Comment by QuantumInformation** [comment]
Does anyone have a minimal example of v7 using typescript? The one [here](https://codesandbox.io/s/m5lxzzpz69) is too complex for my use case. I just want to get up to speed ASAP.

**Comment by QuantumInformation** [comment]
Are there any TS definitions for v6?

**Comment by tannerlinsley** [comment]
Closing in favor of #1252
